pkg/daemon/update.go

```diff
@@ -93,9 +93,56 @@ func getNodeRef(node *corev1.Node) *corev1.ObjectReference {
 	}
 }
 
-// finalizeAndReboot is the last step in an update(), and it can also
-// be called as a special case for the "bootstrap pivot".
-func (dn *Daemon) finalizeAndReboot(newConfig *mcfgv1.MachineConfig) (retErr error) {
+func reloadCrioConfig() error {
+	_, err := runGetOut("pkill", "-HUP", "crio")
+	return err
+}
+
+// performRebootAction takes action based on what rebootAction has been asked.
+// For non-reboot action, it applies configuration, updates node's config and state.
+// In the end uncordon node to schedule workload.
+// If at any point an error occurs, we reboot the node so that node has correct configuration.
+func (dn *Daemon) performRebootAction(action rebootAction, configName string) error {
+	switch action {
+	case rebootActionNone:
+		dn.logSystem("Node has Desired Config %s, skipping reboot", configName)
+	case rebootActionReloadCrio:
+		if err := reloadCrioConfig(); err != nil {
+			dn.logSystem("Reloading crio configuration failed, rebooting: %v", err)
+			dn.reboot(fmt.Sprintf("Node will reboot into config %s", configName))
```

Comment:
One thing I'm debating is whether we want to hard-reboot here or block + degrade. It seems to me that if we are explicitly supporting "rebootless updates" that we should not reboot unexpectedly because of a failure. We should probably do e.g. what we do when a file write fails. This might also allow us not to drain (although recovery is another thing we need to consider) What do you think @sinnykumari ?

Reply:
f you degrade, the customer can't really recover from it and _do_ something to make it work on a retry... so degrading might mean that it can't proceed. not sure how that cuts but since it's an optimization it feels like rebooting should be the fall back.

Reply:
(at least to me rebootless upgrades isn't a promise, it's a best effort optimization in certain cases where it can be safely done)

Reply:
We can but isn't it better to reboot (crio will restart on reboot) and fix the problem by itself rather than degrade? I think we are not promising rebootless update rather it is a best effort basis i.e. MCO tries its best not to reboot for certain config change but if an error occur it reboots. As this is new feature, we can frame documentation accordingly.

Reply:
we can definitely become smarter in future.

Reply:
ha jinx! sinny and i both wrote our responses at the same time :laughing: